File: packages/python/examples/pytest/shadow_dom_test.py
from os import getenv

from pytest import mark


@mark.xfail(
    "appium" in getenv("ALUMNIUM_DRIVER", "selenium"),
    reason="Shadow DOM support is not implemented in Appium yet",
)
def test_shadow_dom(al, navigate):
    navigate("shadow_dom.html")

    page_text = al.get("page text")
    assert "This is inside Shadow DOM!" in page_text
    assert "This is another text inside Shadow DOM!" in page_text

    al.do("click first shadow button")
    page_text = al.get("page text")
    assert "Shadow Button 1 was clicked!" in page_text
    assert "This is inside Shadow DOM!" not in page_text

    al.do("click second shadow button")
    page_text = al.get("page text")
    assert "Shadow Button 2 was clicked!" in page_text
    assert "This is another text inside Shadow DOM!" not in page_text

File: packages/typescript/tests/system/shadowDom.test.ts
import { describe } from "vitest";
import { baseIt } from "./helpers.ts";

describe("Shadow DOM", () => {
  const it = baseIt.override("setup", async ({ setup, skip }) => {
    return async (options) => {
      const result = await setup(options);
      const { isMobile } = result;

      if (isMobile) skip("Shadow DOM support is not implemented in Appium yet");

      return result;
    };
  });

  it("pierces Shadow DOM content", async ({ expect, setup }) => {
    const { al, $ } = await setup();
    await $.navigate("shadow_dom.html");

    let pageText = await al.get("page text");
    expect(pageText).toContain("This is inside Shadow DOM!");
    expect(pageText).toContain("This is another text inside Shadow DOM!");

    await al.do("click first shadow button");
    pageText = await al.get("page text");
    expect(pageText).toContain("Shadow Button 1 was clicked!");
    expect(pageText).not.toContain("This is inside Shadow DOM!");

    await al.do("click second shadow button");
    pageText = await al.get("page text");
    expect(pageText).toContain("Shadow Button 2 was clicked!");
    expect(pageText).not.toContain("This is another text inside Shadow DOM!");
  });
});
